airflow/dag_processing/manager.py

```diff
@@ -576,8 +585,11 @@ def _run_parsing_loop(self):
         while True:
             loop_start_time = time.monotonic()
             ready = multiprocessing.connection.wait(self.waitables.keys(), timeout=poll_time)
-            if self._job:
-                self._job.heartbeat()
+            # we cannot (for now) define job in _job_runner nicely due to circular references of
+            # job and job runner, so we have to use getattr, but we might address it in the future
+            # change when decoupling these two even more
+            if getattr(self._job_runner, "job", None) is not None:
+                perform_heartbeat(self._job_runner.job, only_if_necessary=False)
```

Comment:
I have a refactoring idea but that can wait until after this is merged and doesn’t need to block your other PRs.

Reply:
I think we will not need it at the end -> when we complet the refactoring (final state is #30376), this line is gone, we have job defined individually in each *JobRunner and we always know which runner we access, so there is no need to add the typeguard I think